cdc/processor/sinkmanager/manager_impl.go

```diff
@@ -0,0 +1,183 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package sinkmanager
+
+import (
+	"context"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/pingcap/tiflow/cdc/model"
+	"github.com/pingcap/tiflow/cdc/processor/pipeline"
+	"github.com/pingcap/tiflow/cdc/redo"
+	"github.com/pingcap/tiflow/pkg/sorter"
+)
+
+const (
+	defaultGenerateTaskInterval = 100 * time.Millisecond
+)
+
+// Assert SinkManager implementation
+var _ Manager = (*ManagerImpl)(nil)
+
+// ManagerImpl is the implementation of Manager.
+type ManagerImpl struct {
+	changefeed model.ChangeFeedID
+	// processorCtx is the context of the processor.
+	processorCtx context.Context
+	// progressHeap is the heap of the table progress.
+	progressHeap *tablesFetchProgress
+	// memQuota is used to control the total memory usage of the sink manager.
+	memQuota memQuota
+	// tableSinks is a map from tableID to tableSink.
+	tableSinks sync.Map
+	// sortEngine is used by the sink manager to fetch data.
+	sortEngine sorter.EventSortEngine
+	// redoManager was used to report the resolved ts of the table,
+	// if redo log is enabled.
+	redoManager redo.LogManager
+	// lastBarrierTs is the last barrier ts.
+	lastBarrierTs atomic.Uint64
+	// taskTicker is used to generate fetch tasks.
+	taskTicker *time.Ticker
+	// workers used to pull data from source manager.
+	workers []worker
+	// tableTaskChan is used to send tasks to workers.
+	tableTaskChan chan *tableSinkTask
+	// closedChan is used to notify the manager is closed.
+	closedChan chan struct{}
+}
+
+// New creates a new sink manager.
+func New(redoManager redo.LogManager) Manager {
+	m := &ManagerImpl{
+		progressHeap:  newTableFetchProgress(),
+		tableSinks:    sync.Map{},
+		redoManager:   redoManager,
+		tableTaskChan: make(chan *tableSinkTask),
+		taskTicker:    time.NewTicker(defaultGenerateTaskInterval),
+	}
+
+	// TODO create workers and start them.
+
+	return m
+}
+
+// generateTableSinkFetchTask generates a fetch tableSinkTask to fetch data from the source manager.
+func (m *ManagerImpl) generateTableSinkFetchTask() error {
+	for {
+		select {
+		case <-m.closedChan:
+			return nil
+		case <-m.taskTicker.C:
+			slowestTableProgress := m.progressHeap.pop()
+			tableSink, ok := m.tableSinks.Load(slowestTableProgress.tableID)
+			if !ok {
+				// TODO log it and continue.
+				// Maybe the table sink is removed by the processor.
+				// So we do **not** need add it back to the heap.
+				continue
+			}
+			// If redo log is disabled, we use the barrier ts as the upper bound of the fetch tableSinkTask.
+			// Because can not exceed the barrier ts.
+			barrierTs := m.lastBarrierTs.Load()
+			upperBoundPos := sorter.Position{
+				StartTs:  barrierTs - 1,
+				CommitTs: barrierTs,
+			}
+			canBeAdvance := slowestTableProgress.nextLowerBoundPos.Less(upperBoundPos)
```

Comment:
Seems should use `LessOrEqual`, because upper bound is included.

Reply:
Please check out https://github.com/pingcap/tiflow/pull/7388/commits/9883f4573a646cb7ff784959adadad24ea8d7bf3